### PR TITLE
Allow to set provider value comparer when configuring value conversion

### LINKED ISSUE #27738: Allow to set provider value comparer when configuring value conversion
This would allow to use case-insensitive comparison for example.

Consider showing a warning when an FK or PK property has a custom value comparer but provider value comparer is not specified

### LINKED ISSUE #27791: GetProperty should return null if not found; instead it throws an exception
We have a disconnect between what the documents say should be happening versus what's implemented.

In the file IReadOnlyEntityType.cs when you try to get a property that does not exist it throws an InvalidOperationException whereas the expected return value is null.

Example

```C#
var entityType = MyDbContext.Set<E>().EntityType;
var property = entityType.GetProperty(invalidPropertyName); // this throws an exception
```

### Include stack traces

```
System.InvalidOperationException: The property 'MyTable.invalidPropertyName' could not be found. Ensure that the property exists and has been included in the model.
   at Microsoft.EntityFrameworkCore.Metadata.IReadOnlyEntityType.GetProperty(String name)
   at Microsoft.EntityFrameworkCore.Metadata.IEntityType.GetProperty(String name)
```

### Include provider and version information

EF Core version: 6.0.3
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6
Operating system: Windows 10
IDE: VS 2022

### LINKED ISSUE #27850: Use client value comparer as provider value comparer if possible
This might help reduce the impact of https://github.com/dotnet/EntityFramework.Docs/issues/3801
See also: https://github.com/dotnet/efcore/issues/27762

### PR DESCRIPTION
**Pull request by AndriySvyryd**
- Use client value comparer as provider value comparer if possible
	- The client value comparer will not be used if the model type is a nullable value type since we need the types to match exactly and the provider type is never nullable
- Fix some API issues
- Remove some unused code

Fixes #27738
Fixes #27850
Fixes #27791
Part of #27588

### CODE REVIEW

File: src/EFCore/Metadata/Builders/PropertyBuilder`.cs [hunk 1]
```diff
@@ -440,6 +487,21 @@ public new virtual PropertyBuilder<TProperty> HasConversion<TConversion>(ValueCo
...
         => HasConversion((ValueConverter?)converter, valueComparer);
 
+    /// <summary>
+    ///     Configures the property so that the property value is converted to and from the database
+    ///     using the given <see cref="ValueConverter{TModel,TProvider}" />.
+    /// </summary>
+    /// <typeparam name="TProvider">The store type generated by the converter.</typeparam>
+    /// <param name="converter">The converter to use.</param>
+    /// <param name="valueComparer">The comparer to use for values before conversion.</param>
+    /// <param name="providerComparer">A type that derives from <see cref="ValueComparer" /> to use for the provider values.</param>
```
Comment (ajcvickers) [inline src/EFCore/Metadata/Builders/PropertyBuilder`.cs:497]: Why does only the `providerComparer` doc say it needs to derive from `ValueComparer`? Shouldn't this be consistent?

Also, I've been told in the past that saying "derives from" is ambiguous, and "inherits from" is clearer. But maybe you were all trolling me...

Comment (ajcvickers) [reply src/EFCore/Metadata/Builders/PropertyBuilder`.cs:497]: Also, I guess it's not "a type". Maybe that's why they are inconsistent?

File: src/EFCore/Metadata/IProperty.cs [hunk 2]
```diff
@@ -84,13 +84,17 @@ IEqualityComparer<TProperty> CreateKeyEqualityComparer<TProperty>()
...
     /// </summary>
     /// <returns>The comparer.</returns>
-    [DebuggerStepThrough]
     new ValueComparer GetValueComparer();
 
     /// <summary>
     ///     Gets the <see cref="ValueComparer" /> to use with keys for this property.
     /// </summary>
     /// <returns>The comparer.</returns>
-    [DebuggerStepThrough]
```
Comment (ajcvickers) [inline src/EFCore/Metadata/IProperty.cs:94]: Why no longer step through?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore/Metadata/IProperty.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/IProperty.cs:94]: Doesn't make much sense putting this on the body-less declaration.

**Pull request merged**